compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs

```diff
@@ -773,15 +779,24 @@ impl<'a, 'tcx> InferCtxtExt<'tcx> for InferCtxt<'a, 'tcx> {
                         // }
                         // ```
 
-                        err.span_suggestion(
-                            span,
-                            &format!(
-                                "consider{} borrowing here",
-                                if mtbl { " mutably" } else { "" }
-                            ),
-                            format!("&{}{}", if mtbl { "mut " } else { "" }, snippet),
-                            Applicability::MaybeIncorrect,
-                        );
+                        if imm_result && mut_result {
+                            err.span_suggestions(
+                                span,
+                                "consider borrowing here",
+                                [format!("&{}", snippet), format!("&mut {}", snippet)].into_iter(),
+                                Applicability::MaybeIncorrect,
+                            );
+                        } else {
+                            err.span_suggestion(
+                                span,
+                                &format!(
+                                    "consider{} borrowing here",
+                                    if mut_result { " mutably" } else { "" }
+                                ),
+                                format!("&{}{}", if mut_result { "mut " } else { "" }, snippet),
```

Comment:
This change isn't necessary, but might look better:

```suggestion
                            err.span_suggestion_verbose(
                                span.shrink_to_lo(),
                                &format!(
                                    "consider{} borrowing here",
                                    if mut_result { " mutably" } else { "" }
                                ),
                                format!("&{}", if mut_result { "mut " } else { "" }),
```

I _would_ like us to add a test exercising the `consider mutably borrowing here` case, for which you might have to write a new trait impl that's only valid if it is a mutable borrow (and likely one for the non mutable case as well).